src/include/expression/case_expression.h

```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+//
+//                         PelotonDB
+//
+// case_expression.h
+//
+// Identification: src/include/expression/case_expression.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "expression/abstract_expression.h"
+#include "expression/comparison_expression.h"
+#include "common/sql_node_visitor.h"
+
+namespace peloton {
+namespace expression {
+
+//===----------------------------------------------------------------------===//
+// CaseExpression
+//===----------------------------------------------------------------------===//
+
+class CaseExpression : public AbstractExpression {
+ public:
+  using AbsExprPtr = std::unique_ptr<AbstractExpression>;
+  using WhenClause = std::pair<AbsExprPtr,AbsExprPtr>;
+
+  CaseExpression(type::Type::TypeId type_id, AbsExprPtr argument,
+                 std::vector<WhenClause> &when_clauses,
+                 AbsExprPtr default_expr)
+      : AbstractExpression(ExpressionType::OPERATOR_CASE_EXPR, type_id),
+        argument_(std::move(argument)),
+        clauses_(std::move(when_clauses)),
+        default_expr_(std::move(default_expr)) {
+    // If the argument is provided, then we create When Clauses based on it
+    // Otherwise, we assume that the When Clauses provided are sufficient
+    if (argument_ != nullptr) {
+      for (uint32_t i = 0; i < clauses_.size(); i++)
+		clauses_[i].first.reset(
+           new peloton::expression::ComparisonExpression(
+              peloton::ExpressionType::COMPARE_EQUAL, argument_->Copy(),
+              clauses_[i].first->Copy()));
+    }
+  }
+  type::Value Evaluate(const AbstractTuple *tuple1, const AbstractTuple *tuple2,
```

Comment:
I think we should have two constructors: one with an argument and one without an argument. This makes it's use a little clearer. Thoughts?